### PR TITLE
Select and Dropdown components

### PR DESCRIPTION
**Pull request by tiago**

### CODE REVIEW

File: src/app/dropdown/shared/components/dropdown/dropdown.component.scss [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+$button-height: 32px;
+
+:host {
+  display: block;
+  height: $button-height;
+  position: relative;
+}
+
+:host ::ng-deep keds-tooltip {
```
Comment (frodrigues-kununu) [inline src/app/dropdown/shared/components/dropdown/dropdown.component.scss]: 👀

File: src/app/dropdown/shared/components/dropdown/dropdown.component.spec.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+  });
+
+  beforeEach(() => {
+    fixture = TestBed.createComponent(DropdownComponent);
+    component = fixture.componentInstance;
+
+    fixture.detectChanges();
+  });
+
+  it('should work', () => {
```
Comment (frodrigues-kununu) [inline src/app/dropdown/shared/components/dropdown/dropdown.component.spec.ts]: This might benefit from a better description e.g. `should create a barebones dropdown`.
Edit: nvm, forgot the tests are still in progress.

File: src/app/dropdown/shared/components/multi-select-dropdown/multi-select-dropdown.component.scss [hunk 3]
```diff
@@ -0,0 +1,5 @@
+@import '../../../../core/typography';
+
+.keds-dropdown__label {
+  @extend %typography__static--size-xxs;
```
Comment (frodrigues-kununu) [inline src/app/dropdown/shared/components/multi-select-dropdown/multi-select-dropdown.component.scss]: Why is this needed?

File: src/app/dropdown/shared/components/simple-select-dropdown/simple-select-dropdown.component.ts [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+  @Input('kedsSimpleSelectDropdownItems') items: SimpleSelect.Items = [];
+
+  @Output('kedsSimpleSelectDropdownActiveChange') activeChange: EventEmitter<SimpleSelect.Item> = new EventEmitter();
+
+  expanded: boolean = false;
+
+  ngOnChanges(): void {
+    if (!this.items.includes(this.active)) {
+      this.active = this.items[0];
```
Comment (frodrigues-kununu) [inline src/app/dropdown/shared/components/simple-select-dropdown/simple-select-dropdown.component.ts]: I guess in our case, defaulting to the first item might be best the case like you do here. However, do you think there might be use cases where you can have 0 or 1 items selected in a simple select?

File: src/app/select/shared/components/simple-select/simple-select.component.spec.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
+import { async, ComponentFixture, TestBed } from '@angular/core/testing';
+
+import { SimpleSelectComponent } from './simple-select.component';
+
+describe('SimpleSelectComponent', () => {
+  let component: SimpleSelectComponent;
+  let fixture: ComponentFixture<SimpleSelectComponent>;
+
+  beforeEach(async(() => {
```
Comment (frodrigues-kununu) [inline src/app/select/shared/components/simple-select/simple-select.component.spec.ts]: don't forget to update this with `ng-bullet`

File: src/app/dropdown/shared/components/simple-select-dropdown/simple-select-dropdown.component.ts (hunk 4 above)
Comment (frodrigues-kununu) [reply src/app/dropdown/shared/components/simple-select-dropdown/simple-select-dropdown.component.ts]: Also it sucks a bit that we have to keep this piece of code both here and in `simple-select`, but I can't really think about other way to propagate changes besides this (except maybe for a service which I don't think is justifiable in this case. It would also be a nightmare to manage all the dropdowns and their states in said service).

File: src/app/dropdown/shared/components/dropdown/dropdown.component.scss (hunk 1 above)
Comment (RicardoNeves) [reply src/app/dropdown/shared/components/dropdown/dropdown.component.scss]: This can be fixed if we remove the styles to the global scss file like we do with others components.
If we need to use host, we can always add keds-dropdown class to the host element.

File: src/app/dropdown/shared/components/multi-select-dropdown/multi-select-dropdown.component.scss (hunk 3 above)
Comment (RicardoNeves) [reply src/app/dropdown/shared/components/multi-select-dropdown/multi-select-dropdown.component.scss]: @frodrigues-kununu So that we can reuse keds typography rules.

File: src/app/dropdown/shared/components/dropdown/dropdown.component.scss (hunk 1 above)
Comment (tiago) [reply src/app/dropdown/shared/components/dropdown/dropdown.component.scss]: Looks like a plan 👍. I think I will also need to move the tooltip styles, but that should be easy. Thanks

File: src/app/dropdown/shared/components/simple-select-dropdown/simple-select-dropdown.component.ts (hunk 4 above)
Comment (tiago) [reply src/app/dropdown/shared/components/simple-select-dropdown/simple-select-dropdown.component.ts]: Default to the first one: if we have different requirement it can be changed. No need for that right now.

Output propagation: using a service is overkill.

File: src/app/select/shared/components/simple-select/simple-select.component.spec.ts (hunk 5 above)
Comment (tiago) [reply src/app/select/shared/components/simple-select/simple-select.component.spec.ts]: 👍

File: src/app/select/select.module.ts [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+import { NgModule } from '@angular/core';
+
+import { IconModule } from '../icon/icon.module';
+
+import { MultiSelectComponent } from './shared/components/multi-select/multi-select.component';
+import { SimpleSelectComponent } from './shared/components/simple-select/simple-select.component';
+
+@NgModule({
+  declarations: [MultiSelectComponent, SimpleSelectComponent],
+  exports: [MultiSelectComponent, SimpleSelectComponent],
```
Comment (jmagalhaes-xing) [inline src/app/select/select.module.ts]: Multi line too?

File: src/app/select/shared/components/multi-select/multi-select.component.html [hunk 7]
```diff
@@ -0,0 +1,15 @@
+<div class="keds-multi-select">
+  <div
+    *ngFor="let item of items"
+    (click)="handleItemClick(item)"
```
Comment (jmagalhaes-xing) [inline src/app/select/shared/components/multi-select/multi-select.component.html]: Move the `click` to the end?

File: src/app/select/shared/components/multi-select/multi-select.component.html [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+    class="keds-multi-select__item"
+    [class.keds-multi-select__item--enabled]="!item.disabled"
+    [class.keds-multi-select__item--disabled]="item.disabled"
+    [class.keds-multi-select__item--selected]="item.selected"
+  >
+    <span class="keds-multi-select__checkbox" role="checkbox" [attr.aria-checked]="item.selected">
```
Comment (jmagalhaes-xing) [inline src/app/select/shared/components/multi-select/multi-select.component.html]: Sort attributes?

File: src/app/select/shared/components/multi-select/multi-select.component.spec.ts [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+  let component: MultiSelectComponent;
+  let fixture: ComponentFixture<MultiSelectComponent>;
+
+  const getItems = (): DebugElement[] => fixture.debugElement.queryAll(By.css('.keds-multi-select__item'));
+  const getSelectedItems = (): DebugElement[] => fixture.debugElement.queryAll(By.css('.keds-multi-select__item--selected'));
+
+  configureTestSuite(() => {
+    TestBed.configureTestingModule({
+      declarations: [
+        MultiSelectComponent,
```
Comment (jmagalhaes-xing) [inline src/app/select/shared/components/multi-select/multi-select.component.spec.ts]: Single line?

File: src/app/select/shared/components/multi-select/multi-select.component.spec.ts [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+      ],
+      imports: [
+        MockModule(IconModule),
```
Comment (jmagalhaes-xing) [inline src/app/select/shared/components/multi-select/multi-select.component.spec.ts]: Single line?

File: src/app/select/shared/components/simple-select/simple-select.component.html [hunk 11]
```diff
@@ -0,0 +1,12 @@
+<div class="keds-simple-select">
+  <div
+    *ngFor="let item of items"
+    (click)="handleItemClick(item)"
```
Comment (jmagalhaes-xing) [inline src/app/select/shared/components/simple-select/simple-select.component.html]: Move `click` to the bottom?

File: src/app/select/shared/components/simple-select/simple-select.component.html [hunk 12]
```diff
@@ -0,0 +1,12 @@
...
+    class="keds-simple-select__item"
+    [class.keds-simple-select__item--enabled]="!item.disabled"
```
Comment (jmagalhaes-xing) [inline src/app/select/shared/components/simple-select/simple-select.component.html]: Same here, do we need this?

File: src/app/select/shared/components/multi-select/multi-select.component.html [hunk 13]
```diff
@@ -0,0 +1,15 @@
+<div class="keds-multi-select">
+  <div
+    *ngFor="let item of items"
+    (click)="handleItemClick(item)"
+    class="keds-multi-select__item"
+    [class.keds-multi-select__item--enabled]="!item.disabled"
```
Comment (jmagalhaes-xing) [inline src/app/select/shared/components/multi-select/multi-select.component.html]: Same here, do we need this?

File: src/app/tooltip/_tooltip.scss [hunk 14]
```diff
@@ -1,64 +1,66 @@
...
   padding-right: $tooltip-default-padding + $dismiss-padding;
 }
 
-.tooltip--size-compact.tooltip--dismissible .tooltip__content {
+.keds-tooltip--size-compact.keds-tooltip--dismissible .keds-tooltip__content {
   padding-right: $tooltip-compact-padding + $dismiss-padding;
 }
 
 // TOP Positioning Styles
-:host[x-placement^='top'] .tooltip {
```
Comment (jmagalhaes-xing) [inline src/app/tooltip/_tooltip.scss]: What's the difference here? Between `host` and `keds-tooltip`?

File: src/app/tooltip/shared/components/tooltip/tooltip.component.spec.ts [hunk 15]
```diff
@@ -12,12 +13,15 @@ describe('TooltipComponent', () => {
...
+  const getDismissIcon = (): DebugElement => fixture.debugElement.query(By.directive(IconComponent));
+  const getArrow = (): DebugElement => fixture.debugElement.query(By.css('.keds-tooltip__arrow'));
+
   configureTestSuite(() => {
     TestBed.configureTestingModule({
       declarations: [TooltipComponent],
       imports: [MockModule(IconModule)],
     })
-      .compileComponents();
+      ;
```
Comment (jmagalhaes-xing) [inline src/app/tooltip/shared/components/tooltip/tooltip.component.spec.ts]: Another one here

File: src/app/tooltip/shared/components/tooltip/tooltip.component.ts [hunk 16]
```diff
@@ -5,17 +5,19 @@ import { Core } from '../../../../core/typings';
...
 
-  @Input() theme: Tooltip.Theme;
+  @Input() dismissible: boolean;
 
   @Input() size: Tooltip.Size;
 
-  @Input() dismissible: boolean;
+  @Input() templateRef: TemplateRef<any>;
+
+  @Input() showArrow: boolean;
```
Comment (jmagalhaes-xing) [inline src/app/tooltip/shared/components/tooltip/tooltip.component.ts]: Sort

File: src/app/dropdown/shared/components/dropdown/dropdown.component.html [hunk 17]
```diff
@@ -0,0 +1,32 @@
+<div class="keds-dropdown">
+  <div
+    [attr.aria-disabled]="disabled"
+    role="button"
+    class="keds-dropdown__button"
+    [class.keds-dropdown__button--disabled]="disabled"
+    [class.keds-dropdown__button--enabled]="!disabled"
```
[the remark this answers is not in the document]
Comment (tiago) [reply src/app/dropdown/shared/components/dropdown/dropdown.component.html]: What do you mean by `this`?
All the enabled/disabled/expanded classes? Just to make styling easier.

[the remark this answers is not in the document]
Comment (tiago) [reply src/app/dropdown/shared/components/dropdown/dropdown.component.html]: In particular, `keds-dropdown__button--enabled` is here to avoid doing CSS rules like `keds-dropdown__button:not(keds-dropdown__button--disabled)`

File: src/app/select/shared/components/multi-select/multi-select.component.html (hunk 13 above)
Comment (tiago) [reply src/app/select/shared/components/multi-select/multi-select.component.html]: See comment above. It's here to avoid long CSS queries targeting just "non-disabled" items.

File: src/app/tooltip/_tooltip.scss (hunk 14 above)
Comment (tiago) [reply src/app/tooltip/_tooltip.scss]: The CSS was "removed" from the components. So `:host` became meaningless.
This came out of another comment/thread with @RicardoNeves, but I can't find it right now.

File: src/app/badge/shared/directives/badge/badge.directive.spec.ts [hunk 18]
```diff
@@ -28,7 +28,7 @@ describe('BadgeDirective', () => {
           useClass: Renderer2Mock,
         },
       ],
-    }).compileComponents();
```
[the remark this answers is not in the document]
Comment (tiago) [reply src/app/badge/shared/directives/badge/badge.directive.spec.ts]: Not needed with ng-bullet.

File: src/app/dropdown/shared/components/dropdown/dropdown.component.html (hunk 17 above)
[the remark this answers is not in the document]
Comment (jmagalhaes-xing) [reply src/app/dropdown/shared/components/dropdown/dropdown.component.html]: Yes, I was talking about having the `[class.keds-dropdown__button--enabled]="!disabled"` when you could have the CSS applied to `keds-dropdown__button`
